codex/stores/repostore/operations.nim

```diff
@@ -0,0 +1,213 @@
+): Future[?!void] {.async.} =
+  if cid.isEmpty:
+    return success()
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  await self.metaDs.modify(metaKey,
+    proc (maybeCurrBlockMd: ?BlockMetadata): Future[?BlockMetadata] {.async.} =
+      if currBlockMd =? maybeCurrBlockMd:
+        BlockMetadata(
+          size: currBlockMd.size,
+          expiry: max(currBlockMd.expiry, minExpiry),
+          refCount: currBlockMd.refCount + plusRefCount - minusRefCount
+        ).some
+      else:
+        raise newException(BlockNotFoundError, "Metadata for block with cid " & $cid & " not found")
+  )
+
+proc storeBlock*(self: RepoStore, blk: Block, minExpiry: SecondsSince1970): Future[?!StoreResult] {.async.} =
+  if blk.isEmpty:
+    return success(StoreResult(kind: AlreadyInStore))
+
+  without metaKey =? createBlockExpirationMetadataKey(blk.cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, blk.cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyGet(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, StoreResult)] {.async.} =
+      var
+        md: BlockMetadata
+        res: StoreResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.size == blk.data.len.NBytes:
+          md = BlockMetadata(size: currMd.size, expiry: max(currMd.expiry, minExpiry), refCount: currMd.refCount)
+          res = StoreResult(kind: AlreadyInStore)
+
+          # making sure that the block acutally is stored in the repoDs
+          without hasBlock =? await self.repoDs.has(blkKey), err:
+            raise err
+
+          if not hasBlock:
+            warn "Block metadata is present, but block is absent. Restoring block.", cid = blk.cid
+            if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+              raise err
+        else:
+          raise newException(CatchableError, "Repo already stores a block with the same cid but with a different size, cid: " & $blk.cid)
+      else:
+        md = BlockMetadata(size: blk.data.len.NBytes, expiry: minExpiry, refCount: 0)
+        res = StoreResult(kind: Stored, used: blk.data.len.NBytes)
+        if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+          raise err
+
+      (md.some, res)
+  )
+
+proc tryDeleteBlock*(self: RepoStore, cid: Cid, expiryLimit = SecondsSince1970.low): Future[?!DeleteResult] {.async.} =
+  if cid.isEmpty:
+    return success(DeleteResult(kind: InUse))
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyGet(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, DeleteResult)] {.async.} =
+      var
+        maybeMeta: ?BlockMetadata
+        res: DeleteResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.refCount == 0 or currMd.expiry < expiryLimit:
```

Comment:
is there really a case where we want to delete things even if they have a refcount > 0?

Reply:
Block expiry is such case. This change still assumes that expiry is property of a block, not of a dataset, PR that changes that is here https://github.com/codex-storage/nim-codex/pull/678 In this PR this condition is changed to

```nim
if currMd.refCount == 0:
```